_posts/2021-05-18-My-First-Steps-in-OpenSearch-Plugins.markdown

```diff
@@ -0,0 +1,618 @@
+
+To get started, create a new gradle project. Then, add the following directories and files, which we will edit along the way.
+(Alternatively you can clone the complete source code for this plugin [here](https://github.com/AmiStrn/OpenSearch-Plugin-example-REST))
+```
+.
+|-- LICENSE.txt
+|-- NOTICE.txt
+|-- build.gradle
+`-- src
+    |-- main
+    |   `-- java
+    |       `-- org
+    |           `-- opensearch
+    |               `-- rest
+    |                   `-- action
+    |                       |-- HelloWorldPlugin.java
+    |                       |-- HelloWorldService.java
+    |                       `-- RestHelloWorldAction.java
+    |-- test
+    |   `-- java
+    |       `-- org
+    |           `-- opensearch
+    |               `-- rest
+    |                   `-- action
+    |                       `-- HelloWorldPluginTests.java
+    `-- yamlRestTest
+        |-- java
+        |   `-- org
+        |       `-- opensearch
+        |           `-- rest
+        |               `-- action
+        |                   `-- HelloWorldClientYamlTestSuiteIT.java
+        `-- resources
+            `-- rest-api-spec
+                |-- api
+                |   `-- hello-world.json
+                `-- test
+                    `-- hello-world
+                        |-- 10_basic.yml
+                        `-- 20_hellow_world.yml
+```
+
+---
+## The `build.gradle` of Our Plugin
+All OpenSearch plugins have the same general structure.
+The plugin itself is a zip file that contains at the root:
+- The plugin `jar` files
+- `plugin-descriptor.properties`
+- [OPTIONAL] `bin` directory
+- [OPTIONAL] `config` directory 
+- [OPTIONAL] `plugin-security.policy`
+
+> Any resources a plugin requires are packaged into a resources `jar`. This is due to the fact that only `jar` files at the root are added to the classpath for the plugin.
+
+The `build.gradle` file will build this structure for us when we are done, including creating the required `plugin-descriptor.properties`.
+
+However, since we need the OpenSearch project artifact for the gradle build process and the plugin interface, let's clone the project and publish it to our Maven local. This last step is required since the artifact is not yet published to public repositories. 
+
+In this example I will be using release version `1.0.0-beta1`, as described [here](https://github.com/opensearch-project/opensearch-plugins/blob/main/README.md):
+
+```
+➜  ~ git clone https://github.com/opensearch-project/OpenSearch.git
+➜  ~ cd OpenSearch
+➜  OpenSearch git:(main) git checkout 1.0.0-beta1 -b beta1-release
+➜  OpenSearch git:(beta1-release) ./gradlew publishToMavenLocal -Dbuild.version_qualifier=beta1 -Dbuild.snapshot=false
+```
+
+
+Copy the following to the `build.gradle` file:
+```
+apply plugin: 'java'
+apply plugin: 'opensearch.opensearchplugin'
+apply plugin: 'opensearch.yaml-rest-test'
+
+opensearchplugin {
```

Comment:
It would be nice to explain what `opensearchplugin` is. It would help developers understand what is abstracted.

Reply:
will do. I will keep it very light though since the actual descriptor file takes up a ton of screen space.

Reply:
I do mention that right after the `build.gradle` file is described.

Reply:
Thank you :)